hail/python/hail/methods/statgen.py

```diff
@@ -1509,6 +1509,456 @@ def linear_mixed_regression_rows(entry_expr,
+    ht = ht.group_by(
+        'group'
+    ).aggregate(
+        weight_take = hl.agg.take(ht.weight, n=max_size + 1),
+        G_take = hl.agg.take(ht.G_row, n=max_size + 1),
+        size = hl.agg.count()
+    )
+    ht = ht.annotate(
+        weight = hl.nd.array(hl.or_missing(hl.len(ht.weight_take) <= max_size, ht.weight_take)),
+        G = hl.nd.array(hl.or_missing(hl.len(ht.G_take) <= max_size, ht.G_take)).T
+    )
+    ht = ht.annotate(
+        Q = (((ht.y_residual @ ht.G) * ht.weight) @ ht.G.T) @ ht.y_residual.T
+    )
+    # The paper's linear model explicitly adds an intercept term. We instead require the user to
+    # provide an intercept term, so we set V = \tilde{V} below.
+    #
+    #     N is number of samples
+    #
+    #     V : NxN, P_0: NxN
+    #
+    #     V = s^2 I
+    #     P_0 = V - V X (X.T V X)^{-1} X.T V
+    #     K = G diag(w) G.T
+    #     A = P_0^{1/2} K P_0^{1/2}
+    #
+    # Notice that P_0 is symmetric so it is its own transpose.
+    #
+    # We can rewrite A in terms of Z:
+    #
+    #     Z = P_0^{1/2} G sqrt(diag(w))
+    #     A = Z Z.T
+    #
+    # The non-zero eigenvalues of Z Z.T and Z.T Z are the same so we can instead consider:
+    #
+    #     Z.T Z = sqrt(diag(w)) G.T P_0 G sqrt(diag(w))
+    #
+    # Which we can split into two summands:
+    #
+    #     B = s^2 * sqrt(diag(w)) G.T G sqrt(diag(w))
+    #     C = s^2 * sqrt(diag(w)) G.T (X (X.T X)^{-1} X.T) G sqrt(diag(w))
+    #
+    # We use the reduced QR decomposition to substantially simplify this expression:
+    #
+    #     Q, R = reduced_qr(X)
+    #
+    #     X (X.T X)^{-1} X.T
+    #         = Q R (R.T Q.T Q R)^{-1} R.T Q.T
+    #
+    # Recall that Q.T Q is 1 for the reduced QR:
+    #
+    #         = Q R (R.T R)^{-1} R.T Q.T
+    #
+    # Notice that R is invertible iff R.T R is invertible, so we can push the inversion through the
+    # product
+    #
+    #         = Q R R^{-1} R.T^{-1} R.T Q.T
+    #         = Q R R^{-1} Q.T
+    #         = Q Q.T
+    #
+    # Notice that this is *not* 1 because we're using the reduced QR. Returning to our matrices
+    # above:
+    #
+    #     B = s^2 * sqrt(diag(w)) G.T G sqrt(diag(w))
+    #     C = s^2 * sqrt(diag(w)) G.T Q Q.T G sqrt(diag(w))
+    #
+    # which we can rewrite as Gram matrices :
+    #
+    #     B0 = G sqrt(diag(w))
+    #     C0 = Q.T G sqrt(diag(w))
+    #
+    #     B = s^2 B0.T B0
+    #     C = s^2 C0.T C0
+    #
+    # Back to Z:
+    #
+    #     Z.T Z = s^2 (B0.T B0 - C0.T C0)
+    #
+    # Which we want the eigenvalues of. Hail lacks an eigenvalues method, but luckily the singular
+    # values of a symmetric semi-positive-definite matrix *are* the eigenvalues.
+    #
+    # We avoid the square root in order to avoid complex numbers.
+
+    Q, _ = hl.nd.qr(ht.covmat)
+    C0 = Q.T @ ht.G
+    singular_values = hl.nd.svd(ht.G.T @ (ht.G * ht.weight) - C0.T @ (C0 * ht.weight), compute_uv=False)
```

Comment:
As discussed in zulip, I think a better way to compute this is:
```
R = nd.qr(ht.G - Q @ (Q.T @ G), mode="r")
singular_values = hl.nd.svd(R, compute_uv=False)
eigenvalues = singular_values.map(lambda x: x**2)
```
(replace `singular_values` with `eigenvalues` below)